Mpm/linear implicit

peterjwilson opened the pull request:
- Added linear implicit option for MPM analysis. This skips the computation and addition of the geometric stiffness matrix and does not check for convergence (as is done for structural mechanics). In project parameters specify 'analysis_type = linear' to enable. Nightly test added. The intent of linear implicit is to speed up computation where a linear analysis is sufficiently accurate.

- Added FAQs page for particle mechanics with tips for common problems (that arise mainly due to GiD's Kratos being old). Hopefully this helps other people in the future.

PeterJosephWilson committed:
add linear implicit solver, not yet tested

PeterJosephWilson committed:
Merge branch 'master' into MPM/linear_implicit

PeterJosephWilson committed:
Merge branch 'master' into MPM/linear_implicit

PeterJosephWilson committed:
change name. isolate bugged tests.

PeterJosephWilson committed:
added linear test.

system matrices aren't set to the right size and an error occurs when they are cleared.

peterjwilson committed:
Create FAQs.md

peterjwilson committed:
Update README.md

peterjwilson committed:
Update FAQs.md

PeterJosephWilson committed:
Update mpm_residual_based_linear_strategy.h

PeterJosephWilson committed:
cleanup. add in linear implicit test.

peterjwilson committed:
Update FAQs.md

peterjwilson committed:
extended linear function to other elements

philbucher reviewed (commented):
The linear strategy seems like much copy paste from the core
Cannot you directly use that one?

peterjwilson committed:
Using core linear strategy

As per https://github.com/KratosMultiphysics/Kratos/pull/6401#pullrequestreview-357413932

peterjwilson committed:
Update add_custom_strategies_to_python.cpp

peterjwilson commented:
Good point, that works fine. Thanks Phil :)

tteschemacher commented on applications/ParticleMechanicsApplication/particle_mechanics_application_variables.cpp:
```diff
@@ -90,4 +90,6 @@ namespace Kratos
     KRATOS_CREATE_3D_VARIABLE_WITH_COMPONENTS( NODAL_INERTIA)
 
+    // Solver related variables
+    KRATOS_CREATE_VARIABLE(bool, IMPLICIT_IS_LINEAR)
```
Does the Variable need to involve IMPLICIT? Isn't GEOMETRICAL_LINEAR enough?

peterjwilson replied:
Updated

peterjwilson committed:
Changed bool to IGNORE_GEOMETRIC_STIFFNESS

tteschemacher reviewed (dismissed):
To me it looks fine now

bodhinandach commented on applications/ParticleMechanicsApplication/tests/test_ParticleMechanicsApplication.py:
```diff
@@ -86,13 +94,16 @@ def AssembleTestSuites():
     ### Create a test suit that contains all the tests:
     allSuite = suites['all']
-    allSuite.addTests(nightSuite) # already contains the smallSuite
+    allSuite.addTests(smallSuite)
```
Can you revert this? I think it's better to put the smallsuite in nightly too so that we can get noticed by the community if our unit tests are failing. The nightsuite will also include the smallsuite if you do that.

philbucher replied:
I agree

peterjwilson replied:
Good point

bodhinandach reviewed (changes requested):
I have one minor request. But overall, looks good. Thanks Peter!

peterjwilson committed:
Revert back to including smallsuite in nightsuite

philbucher commented:
do you still need this branch @peterjwilson ?
